### ISSUE TITLE
bias_init  questions

### ISSUE THREAD

**Issue by powermano** [ultralytics/yolov5#1010]
## ❔Question
why the bias of detect modelu need specific init?  Actually, it really works, if  i do not use this bias init(code is below),  the object loss is higher. In my experiments(i reproduce  the code in mxnet,  i change the yolov5s.pt to  onnx, and then change onnx to .params, the test mAP is very close),

mxnet version:
```
Epoch[269] Batch[800]   Yolov5ClassLoss=0.030,  Yolov5ObjLoss=0.083,  Yolov5BoxLoss=0.047 
```
the Yolov5ObjLoss is still very high, and the mAP is about :
Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 23.6

but when i use the  bias init, the objetc loss becomes lower, even in the first warm_up stage:
```
Epoch[1] Batch[400]  Yolov5ClassLoss=0.130,  Yolov5ObjLoss=0.065,  Yolov5BoxLoss=0.127
```

the code is still running and i will show the test mAP later.

I really want to know the intuition of the bias init  and why it  makes a great influence in object loss.

Thanks.

## Additional context
the bias init code is as below:
```
    def _initialize_biases(self, cf=None):  # initialize biases into Detect(), cf is class frequency
        # cf = torch.bincount(torch.tensor(np.concatenate(dataset.labels, 0)[:, 0]).long(), minlength=nc) + 1.
        m = self.model[-1]  # Detect() module
        for f, s in zip(m.f, m.stride):  #  from
            mi = self.model[f % m.i]
            b = mi.bias.view(m.na, -1)  # conv.bias(255) to (3,85)
            b[:, 4] += math.log(8 / (640 / s) ** 2)  # obj (8 objects per 640 image)
            b[:, 5:] += math.log(0.6 / (m.nc - 0.99)) if cf is None else torch.log(cf / cf.sum())  # cls
            mi.bias = torch.nn.Parameter(b.view(-1), requires_grad=True)
```

**Comment by glenn-jocher** [comment]
@powermano the reasoning is simply to set them at what we believe their steady state values may trend to as epochs trend towards infinity. See https://arxiv.org/pdf/1708.02002.pdf section 3.3

**Comment by powermano** [comment]
> @powermano the reasoning is simply to set them at what we believe their steady state values may trend to as epochs trend towards infinity. See https://arxiv.org/pdf/1708.02002.pdf section 3.3

thanks, I have another question about the bias warm lr:
```
 # bias lr falls from 0.1 to lr0, all other lrs rise from 0.0 to lr0
                    x['lr'] = np.interp(ni, xi, [hyp['warmup_bias_lr'] if j == 2 else 0.0, x['initial_lr'] * lf(epoch)])
                    if 'momentum' in x:
                        x['momentum'] = np.interp(ni, xi, [hyp['warmup_momentum'], hyp['momentum']])
```
bias lr falls from 0.1 to lr0, i want to know the intuition of the change of bias lr.  

if the bias lr falls frem 0 to lr0, i got the following  phenomenon：

As mentioned above,  using the bias init actually
 can get object loss around 0.06 in warm up stage,  but it increase from 0,06 to 0.09 in warnup stage,  and then decrease from 0.09 to 0.084.   

I would like to know whether using bias lr falling from 0.1 to lr0 in warmup stage will reduce the loss to a stable and normal state(around 0.06), and then optimize object loss in the later iters. Or there is a problem with my object loss implementation。

**Comment by glenn-jocher** [comment]
@powermano you're free to vary this hyperparameter in your hyp yaml file located in /data.

**Comment by powermano** [comment]
> @powermano you're free to vary this hyperparameter in your hyp yaml file located in /data.
OK
I use yolov5s.yaml config , but i met a serious problem,  the object loss can't  be optimized into a raletively lower status.  After 300 epoch, it is still around 0.08,  which is inconsistent with your official loss changes. 

I am so confused.

**Comment by powermano** [comment]
@glenn-jocher 

Finally, i found that the EMA plays a big role in whole training iters.  
Following is the results of mAP comparison:

| epoch |  without EMA | EMA  |
|---|---|---|
| 10       |       12.82  |  18.11
| 90      |        21.81  |29.42


But i find another problem, my cls and giou loss can not quickly reach a stable point。
![image](https://user-images.githubusercontent.com/24365567/94505903-bef6b800-023e-11eb-8c69-51b273c43780.png)
Your cls and giou loss can be optimized to 0.04 and 0.06 respectively in the first few epochs(As shown in above picture).  But my cls and giou loss can only optimized to 0.86 and 0.73 respectively in the first few epochs. 

I want to know whether this situation is caused by initialization , optimizer or others.

**Comment by glenn-jocher** [comment]
@powermano yes, EMA helps in early training especially, less so after many epochs.

You can not compare two different dataset loss magnitudes, this is like saying your apples are not as good as your neighbors oranges.

**Comment by konioy** [comment]
> 

I have the same question, why "bias lr falls from 0.1 to lr0, all other lrs rise from 0.0 to lr0"?

**Comment by glenn-jocher** [comment]
@konioy this is a strategy I came up with early on since bias is more stable than weights and can tolerate higher LRs without losing stability, but it's been a while since I experimented with warmup strategies, so if you find a better warmup implementation in your experiments please let us know!
